persistence/sql/batch/create.go

```diff
@@ -0,0 +1,139 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package batch
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/gobuffalo/pop/v6"
+	"github.com/gofrs/uuid"
+	"github.com/jmoiron/sqlx/reflectx"
+	"github.com/pkg/errors"
+
+	"github.com/ory/x/otelx"
+	"github.com/ory/x/sqlcon"
+
+	"github.com/ory/x/sqlxx"
+)
+
+type (
+	insertQueryArgs struct {
+		TableName    string
+		ColumnsDecl  string
+		Columns      []string
+		Placeholders string
+	}
+	quoter interface {
+		Quote(key string) string
+	}
+	TracerConnection struct {
+		Tracer     *otelx.Tracer
+		Connection *pop.Connection
+	}
+)
+
+func buildInsertQueryArgs[T any](ctx context.Context, quoter quoter, models []*T) insertQueryArgs {
+	var (
+		v     T
+		model = pop.NewModel(v, ctx)
+
+		columns        []string
+		quotedColumns  []string
+		placeholders   []string
+		placeholderRow []string
+	)
+
+	for _, col := range model.Columns().Cols {
+		quotedColumns = append(quotedColumns, quoter.Quote(col.Name))
+		columns = append(columns, col.Name)
+		placeholderRow = append(placeholderRow, "?")
+	}
+	for range models {
+		placeholders = append(placeholders, fmt.Sprintf("(%s)", strings.Join(placeholderRow, ", ")))
+	}
+
+	return insertQueryArgs{
+		TableName:    quoter.Quote(model.TableName()),
+		ColumnsDecl:  strings.Join(quotedColumns, ", "),
+		Columns:      columns,
+		Placeholders: strings.Join(placeholders, ",\n"),
+	}
+}
+
+func buildInsertQueryValues[T any](mapper *reflectx.Mapper, columns []string, models []*T) (values []any, err error) {
+	now := time.Now().Truncate(time.Microsecond)
+
+	for _, m := range models {
+		m := reflect.ValueOf(m)
+
+		// Append model fields to args
+		for _, c := range columns {
+			field := mapper.FieldByName(m, c)
+
+			switch c {
+			case "created_at", "updated_at":
+				field.Set(reflect.ValueOf(now))
+
+			case "id":
+				if field.Interface().(uuid.UUID) != uuid.Nil {
+					break // breaks switch, not for
+				}
+				id, err := uuid.NewV4()
+				if err != nil {
+					return nil, err
+				}
+				field.Set(reflect.ValueOf(id))
+			}
+			values = append(values, field.Interface())
+
+			// Special-handling for *sqlxx.NullTime: mapper.FieldByName sets this to a zero time.Time,
+			// but we want a nil pointer instead.
+			if i, ok := field.Interface().(*sqlxx.NullTime); ok {
```

Comment:
Could this be

```
if i, ok := field.Interface().(sqlxx.NullTime); ok {
```

as well? Or will it always be a pointer. Example:

```
type foo struct {
  a *sqlxx.NullTime
  b *sqlxx.NullTime
}
```

Reply:
If it is not a pointer to `sqlxx.NullTime` we won't need to correct anything.